### ISSUE TITLE
Support rebase --onto option

### ISSUE THREAD

**Issue by tremby** [pcottle/learnGitBranching#532]
I have to look this one up each time, since I can never remember which branch refs will change, and what will happen to HEAD. It'd be nice to visualize it in this tool.

**Comment by zartc** [comment]
I also think that not having rebase --onto is a big drawback. It would be so nice to be able to demonstrate the difference between a rebase and a rebase --onto.
I would help if only I knew how to develop a javascript application. 
Please can someone take on this task?

**Comment by pcottle** [comment]
yeah I'd love to review a PR for this functionality! I've never used `--onto` before but is it basically the same as mercurial's `-s` parameter? where you specify the source rather than it being assumed to be `HEAD`?

If it's that simple I can probably add it

**Comment by tremby** [comment]
I don't know Mercurial.

Relevant excerpt from `man git-rebase`:

```
       Here is how you would transplant a topic branch based on one branch to
       another, to pretend that you forked the topic branch from the latter
       branch, using rebase --onto.

       First let’s assume your topic is based on branch next. For example, a
       feature developed in topic depends on some functionality which is found
       in next.

               o---o---o---o---o  master
                    \
                     o---o---o---o---o  next
                                      \
                                       o---o---o  topic

       We want to make topic forked from branch master; for example, because
       the functionality on which topic depends was merged into the more
       stable master branch. We want our tree to look like this:

               o---o---o---o---o  master
                   |            \
                   |             o'--o'--o'  topic
                    \
                     o---o---o---o---o  next

       We can get this using the following command:

           git rebase --onto master next topic

       Another example of --onto option is to rebase part of a branch. If we
       have the following situation:

                                       H---I---J topicB
                                      /
                             E---F---G  topicA
                            /
               A---B---C---D  master

       then the command

           git rebase --onto master topicA topicB

       would result in:

                            H'--I'--J'  topicB
                           /
                           | E---F---G  topicA
                           |/
               A---B---C---D  master

       This is useful when topicB does not depend on topicA.
```

**Comment by zartc** [comment]
The canonical forme of the rebase command is:  

    git rebase [--onto <new-base>] <upstream> [<branch>]

Where:  

    --onto <new-base>
        new-base is the commit onto which the new commits will be applied one-by-one.

    <upstream>
        The commit marking the start of the range to rebase.
        It is often the name of the parent branch of the branch we want top rebase (whence the name: upstream), but it can be any commit SHA1.

    <branch>
        The commit marking the end of the range to rebase. It is often the current branch name, but it can be any commit SHA1.
        If not specified it takes the value HEAD by default.


The operation performed internally by rebase --onto are:
1. Switch to `<branch>` or stay on the current branch if `<branch>` is not specified.
2. Copy all commits in the range `<upstream>..<branch>` that is the commits on `<branch>` up to but not including those on `<upstream>`.  
These are the same commits that would be shown by `git log <upstream>..<branch>`
3. The current branch is reset to `<upstream>`, or `<newbase>` if the `--onto` option was supplied.  
This has the exact same effect as `git reset --hard <upstream>` (or `<newbase>`).  
IMPORTANT: `ORIG_HEAD` is set to point at the tip of the branch before the reset.
4. The commits that were previously copied are then reapplied to the current branch, one by one, in order.

It would be nice if the program was able to manage the `ORIG_HEAD` variable. That would allow the user to later cancel the rebase via a `git reset --hard ORIG_HEAD`.

**Comment by pcottle** [comment]
after reading:
https://stackoverflow.com/questions/29914052/i-cant-understand-the-behaviour-of-git-rebase-onto

this is actually pretty complex to support...

The one argument version is simple, it is basically an alias. So this command:
`git rebase <dest> HEAD` becomes `git rebase --onto <dest>`

The two argument version is trickier, since we want to actually *stop* finding upstream commits at the specified ref. and then the three argument version is even more nuts -- we want to start upstream of our current HEAD.

getting all of that to work with arg validation / error states / tests is going to be a lot. I'll leave this open for now if anyone brave wants to try it, but probably going to defer for now.

*(I'd also argue that a lesson on this might be more confusing than helpful if you are just learning git. for expert users it would be helpful though)*

**Comment by zartc** [comment]
The SO responses might make you think that the `rebase --onto` command is terribly complex, but in reality things are much simpler than that.

First of all, there is no such thing as the one or two arguments rebase. Only the three arguments rebase exists. In the one or two arguments scenario some arguments take default values, that's all. This should considerably simplify the understanding of the rebase command.

Second, the individual steps performed by the rebase command are easy to follow and can be summarized as follow:

## preparation

1. If `<branch>` is not specified   =>   let `<branch>` be equal to `HEAD`

2. If `--onto` option and `<new-base>` parameter are not specified  => let `<new-base>` be equal to the `<upstream>` parameter

3. `<upstream>` parameter must be specified,  if not  => error!

## excecution

4. Switch to `<branch>`    (might be a NOP when `<branch>` == `HEAD`)

5. Cut all commits in the range `<upstream>..<branch>`.  That is all commits starting from `<branch>` up-to but not including the shared commits on `<upstream>`. These are the same commits that would be shown by `git log <upstream>..<branch>`

6. Reset the current branch to `<new-base>` as with a `git reset --hard <new-base>`

7. Paste the commits to the current branch, one-by-one.


This algorithm should cover all rebase use-cases. The difficulty, as I see things, is to locate the commits in the range `<upstream>..<branch>` where both `upstream` and `branch` can be expressed as branch name or as commit SHA1 and can be both from the same branch or from different branches. But this might already be implemented in the program, I don't know.

Other than that, it seems that the program is already able to perform each and every individual steps, isn't it ?  (except the management of the ORIG_HEAD variable)

**Comment by pcottle** [comment]
Thanks for the clarification. I guess I meant that all the functionality exists already when the arguments *take on their default values*. Individually sure, I have utility methods to perform almost all of these operations (resolve a ref, get commits between two refs, etc) but linking it all together (with the async animation) just involves more code and testing.

Not saying I won't ever do this but not as simple of a task as I had hoped at first :P

**Comment by pcottle** [comment]
(closing out old issues, also this is super hard)

**Comment by waldyrious** [comment]
For the record, this has now been implemented, in #875! :tada: :tada: 

Thanks, @Qusijue! :heart:
